asdf/commands/edit.py

```diff
@@ -130,23 +131,42 @@ def write_edited_yaml_larger(path, new_content, version):
             pad_length = util.calculate_padding(len(new_content), True, fd.block_size)
             fd.fast_forward(pad_length)
 
+            # now copy over ASDF block contents
+
             with generic_io.get_file(path) as original_fd:
-                # Consume the file up to the first block, which must exist
-                # as a precondition to using this method.
-                original_fd.seek_until(
-                    constants.BLOCK_MAGIC,
-                    len(constants.BLOCK_MAGIC),
-                )
-                ctx = AsdfFile(version=version)
-                blocks = BlockManager(ctx, copy_arrays=False, lazy_load=False)
-                blocks.read_internal_blocks(original_fd, past_magic=True, validate_checksums=False)
-                blocks.finish_reading_internal_blocks()
-                blocks.write_internal_blocks_serial(fd)
-                blocks.write_block_index(fd, ctx)
-                blocks.close()
-
-            # the file needs to be closed here to release all memmaps
-            original_fd.close()
+                original_fd.seek_until(constants.BLOCK_MAGIC, len(constants.BLOCK_MAGIC))
+                old_first_block_offset = original_fd.tell() - len(constants.BLOCK_MAGIC)
+                new_first_block_offset = fd.tell()
+
+                # check if the original file has a block index which we will need to update
+                # as we're moving the blocks
+                block_index_offset = bio.find_block_index(original_fd)
+                if block_index_offset is None:
+                    block_index = None
+                    original_fd.seek(0, generic_io.SEEK_END)
+                    blocks_end = original_fd.tell()
+                else:
+                    blocks_end = block_index_offset
+                    try:
+                        block_index = bio.read_block_index(original_fd, block_index_offset)
+                    except BlockIndexError:
+                        # the original index was invalid
+                        block_index = None
+
+                # copy over blocks byte-for-byte from old_first_block_offset to block_index_offset
+                original_fd.seek(old_first_block_offset)
+                block_index_offset
+                block_size = min(fd.block_size, original_fd.block_size)
+                n_bytes = blocks_end - old_first_block_offset
+                for offset in range(0, n_bytes, block_size):
+                    this_size = min(block_size, n_bytes - offset)
+                    fd.write(original_fd.read(this_size))
+
+                # update index
```

Comment:
I did find that a bit odd that `asdftool edit` would add an index to a file that didn't previously had one (unless it has a streaming block) only if the tree expanded into the blocks. I don't quite recall if at the time I considered this a 'bug' in edit and changed it to only add an index if the original file had one. I would argue that this makes more sense then sometimes adding an index to a file that didn't previously have one.

It shouldn't be a problem to change it back to having it always generate an index. However this would require parsing the last block header to make sure an index isn't added for a file with a streaming block.

Shall we consider the adding of a block index a bug and keep the current behavior in this PR or should I update it to parse the last block header and add an index to a file if the tree expands into the blocks?